third_party/move/move-vm/types/src/resolver.rs

```diff
@@ -44,44 +42,16 @@ pub fn resource_size(resource: &Option<Bytes>) -> usize {
 ///                       are always structurally valid)
 ///                    - storage encounters internal error
 pub trait ResourceResolver {
-    type Error: Debug;
-
     fn get_resource_bytes_with_metadata_and_layout(
         &self,
         address: &AccountAddress,
         struct_tag: &StructTag,
         metadata: &[Metadata],
         layout: Option<&MoveTypeLayout>,
-    ) -> Result<(Option<Bytes>, usize), Self::Error>;
+    ) -> PartialVMResult<(Option<Bytes>, usize)>;
 }
 
 /// A persistent storage implementation that can resolve both resources and modules
-pub trait MoveResolver<E: Debug>: ModuleResolver<Error = E> + ResourceResolver<Error = E> {
-    fn get_resource(
```

Comment:
Any particular reason you are removing these wrappers? I was looking at the callsites affected by this and they now need to deal with extra things they don't seem to care about.

Reply:
Yes! These wrappers create coupling between code and data (which have different lifetimes) - `MoveResolver`. Right now it is stored inside the session, and it makes sense for resources (though it results in fun respawned session code), but not for modules - module storage should be passed by reference to VM APIs: right now it is inside transaction data cache, which is not a right place for it.

Essentially, creating a new session can capture `ResourceResolver` reference inside transaction data cache, but `ModuleResolver` (or a modification of this trait) will passed as a reference to it. This allows us to nicely handle init_module on AptosVM side: we create a version of deltaStorage for modules, and pass its reference when executing init_module functions.

Reply:
Basically, `ModuleResolver` --> `ModuleStorage` and passed to `execute_...` or `load_...` APIs, and where we have `MoveResolver` having `ResourceResolver` sufficies.

Reply:
One option to avoid boilerplate: get_resource/get_resource_with_metadata should take extra argument which allows them to fetch modules.

Reply:
I certainly like more explicitness about what's happening to get the resource, and not having module metadata reads hidden there.